### PR TITLE
En 4248 compact miniblocks

### PR DESCRIPTION
**Pull request by bogdan-rosianu**

### CODE REVIEW

File: process/block/preprocess/transactions.go [hunk 1]
```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
 
+	compressedMiniBlocks := txs.compressMiniBlocks(miniBlocks)
+	if len(compressedMiniBlocks) > 0 {
+		return compressedMiniBlocks, nil
+	}
+
 	return miniBlocks, nil
 }
 
+func (txs *transactions) compressMiniBlocks(miniBlocks block.MiniBlockSlice) block.MiniBlockSlice {
```
Comment (iulianpascalau) [inline process/block/preprocess/transactions.go]: Extract all these methods in a new struct! 
You added around 200 lines of code here.

File: process/block/preprocess/transactions.go [hunk 2]
```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
+		compressedMbs = txs.merge(compressedMbs, miniBlocks[i])
+	}
+
+	log.Info(fmt.Sprintf("compressed from %d miniblocks to %d miniblocks", len(miniBlocks), len(compressedMbs)))
+
+	return compressedMbs
+}
+
+func (txs *transactions) merge(mergedMiniBlocks block.MiniBlockSlice, miniBlock *block.MiniBlock) block.MiniBlockSlice {
+	// todo: check gas limit
```
Comment (iulianpascalau) [inline process/block/preprocess/transactions.go]: Do not forget to implement the check for gas limit

File: process/block/preprocess/transactions.go (hunk 1 above)
Comment (sasurobert) [reply process/block/preprocess/transactions.go]: +1

File: process/block/preprocess/transactions.go (hunk 2 above)
Comment (sasurobert) [reply process/block/preprocess/transactions.go]: this is the most important part. add it before merging.

File: process/block/preprocess/transactions.go [hunk 3]
```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
+				continue
+			}
+			txNfo.txHash = txHash
+			txsInfo[miniBlock.ReceiverShardID] = append(txsInfo[miniBlock.ReceiverShardID], txNfo)
+		}
+	}
+
+	for {
+		foundSomething := false
+		mbSlice, foundSomething = txs.tryCreateMiniBlocks(txsInfo, mapSenderNonce, mbSlice, foundSomething)
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: rename variables and function name. be more specific of what it does.

File: process/block/preprocess/transactions.go [hunk 4]
```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
+	mb.SenderShardID = txs.shardCoordinator.SelfId()
+	mb.TxHashes = make([][]byte, 0)
+	mb.Type = block.TxBlock
+
+	for {
+		found := false
+		for _, tx := range currentTxs {
+			mb.ReceiverShardID = tx.receiverShardID
+			nonce, ok := mapSenderNonce[string(tx.tx.GetSndAddress())]
+			if !ok || tx.tx.GetNonce() == nonce+1 {
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: tx.tx does not look good. rename variables.

File: process/block/preprocess/transactions.go [hunk 5]
```diff
@@ -471,9 +473,175 @@ func (txs *transactions) CreateAndProcessMiniBlocks(
...
+}
+
+func removeItem(currentTxs []*txInfo, tx *txInfo) []*txInfo {
+	for ct := 0; ct < len(currentTxs); ct++ {
+		if currentTxs[ct] == tx {
+			if ct == len(currentTxs)-1 {
+				return currentTxs[:ct]
+			}
+
+			return append(currentTxs[:ct], currentTxs[ct+1:]...)
```
Comment (sasurobert) [inline process/block/preprocess/transactions.go]: this operation has a lot of cost. Refactor.

Comment (sasurobert) [reply process/block/preprocess/transactions.go]: use maps, get all final values into a slice and sort.

File: process/block/preprocess/transactions.go (hunk 1 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactions.go]: Refactored

File: process/block/preprocess/transactions.go (hunk 4 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactions.go]: Refactored

File: process/block/preprocess/transactions.go (hunk 5 above)
Comment (SebastianMarian) [reply process/block/preprocess/transactions.go]: Refactored

File: process/block/preprocess/miniBlocksCompaction.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+type miniBlocksCompaction struct {
+	economicsFee     process.FeeHandler
+	shardCoordinator sharding.Coordinator
+
+	mapHashesAndTxs         map[string]data.TransactionHandler
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: mapHashToTx

File: process/block/preprocess/miniBlocksCompaction.go [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+	mbc.mapSenderNonce = make(map[string]uint64)
+	mbc.mapUnallocatedTxsHashes = make(map[string]struct{})
+
+	return &mbc, nil
+}
+
+// Compact method tries to compact the given mini blocks to have only one mini block per sender/received pair
+func (mbc *miniBlocksCompaction) Compact(
+	miniBlocks block.MiniBlockSlice,
+	mapHashesAndTxs map[string]data.TransactionHandler,
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: mapHashToTx

File: process/block/preprocess/miniBlocksCompaction.go [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+	for index, miniBlock := range miniBlocks {
+		if index == 0 {
+			continue
+		}
+
+		compactedMiniBlocks = mbc.merge(compactedMiniBlocks, miniBlock)
+	}
+
+	if len(miniBlocks) > len(compactedMiniBlocks) {
+		log.Info(fmt.Sprintf("compacted from %d miniblocks to %d miniblocks\n",
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: delete "from"

File: process/block/preprocess/miniBlocksCompaction.go [hunk 9]
```diff
@@ -0,0 +1,289 @@
...
+) block.MiniBlockSlice {
+
+	for _, mergedMiniBlock := range mergedMiniBlocks {
+		sameType := miniBlock.Type == mergedMiniBlock.Type
+		sameSenderShard := miniBlock.SenderShardID == mergedMiniBlock.SenderShardID
+		sameReceiverShard := miniBlock.ReceiverShardID == mergedMiniBlock.ReceiverShardID
+
+		canMerge := sameSenderShard && sameReceiverShard && sameType
+		if canMerge {
+			haveEnoughGasToMerge := mbc.haveEnoughGasToMerge(mergedMiniBlock, miniBlock)
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: it is not about if there is enough gas. Better name: isThereStillEnoughSpace

File: process/block/preprocess/miniBlocksCompaction.go [hunk 10]
```diff
@@ -0,0 +1,289 @@
...
+			}
+		}
+	}
+
+	mergedMiniBlocks = append(mergedMiniBlocks, miniBlock)
+
+	return mergedMiniBlocks
+}
+
+func (mbc *miniBlocksCompaction) haveEnoughGasToMerge(
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: rename function as well.

File: process/block/preprocess/miniBlocksCompaction.go [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+		log.Info(err.Error())
+		return false
+	}
+
+	haveEnoughGasToMerge := gasUsedInDestMiniBlock+gasUsedInSrcMiniBlock <= process.MaxGasLimitPerMiniBlock
+
+	return haveEnoughGasToMerge
+}
+
+func (mbc *miniBlocksCompaction) getGasUsedInMiniBlock(miniBlock *block.MiniBlock) (uint64, error) {
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: calculateUsedGasInMiniblock

File: process/block/preprocess/miniBlocksCompaction.go [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+	return mergedMiniBlocks
+}
+
+func (mbc *miniBlocksCompaction) haveEnoughGasToMerge(
+	destMiniBlock *block.MiniBlock,
+	srcMiniBlock *block.MiniBlock,
+) bool {
+
+	gasUsedInDestMiniBlock, err := mbc.getGasUsedInMiniBlock(destMiniBlock)
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: cache the results for the already calculated miniblocks. You are iterating over and over on the same set of data.

File: process/block/preprocess/miniBlocksCompaction.go [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+	for _, miniBlock := range miniBlocks {
+		for _, txHash := range miniBlock.TxHashes {
+			tx, ok := mbc.mapHashesAndTxs[string(txHash)]
+			if !ok {
+				return nil, process.ErrMissingTransaction
+			}
+
+			nonce, ok := mbc.mapSenderNonce[string(tx.GetSndAddress())]
+			if !ok || nonce > tx.GetNonce() {
+				mbc.mapSenderNonce[string(tx.GetSndAddress())] = tx.GetNonce()
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: better name: minSenderNonce ?

File: process/block/preprocess/miniBlocksCompaction.go [hunk 14]
```diff
@@ -0,0 +1,289 @@
...
+			}
+
+			mbc.mapUnallocatedTxsHashes[string(txHash)] = struct{}{}
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: so initially every transaction is unallocated ? in terms of expansion

File: process/block/preprocess/miniBlocksCompaction.go (hunk 9 above)
Comment (SebastianMarian) [reply process/block/preprocess/miniBlocksCompaction.go]: I will rename in isEnoughGasSpace

File: process/block/preprocess/miniBlocksCompaction.go (hunk 10 above)
Comment (SebastianMarian) [reply process/block/preprocess/miniBlocksCompaction.go]: I will rename in isEnoughGasSpace

File: process/block/preprocess/miniBlocksCompaction.go (hunk 14 above)
Comment (SebastianMarian) [reply process/block/preprocess/miniBlocksCompaction.go]: yes

File: process/block/preprocess/miniBlocksCompaction_test.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+	"github.com/ElrondNetwork/elrond-go/data/block"
+	"github.com/ElrondNetwork/elrond-go/data/transaction"
+	"github.com/ElrondNetwork/elrond-go/dataRetriever/shardedData"
+	"github.com/ElrondNetwork/elrond-go/process"
+	"github.com/ElrondNetwork/elrond-go/process/mock"
+	"github.com/ElrondNetwork/elrond-go/storage/storageUnit"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMiniBlocksCompaction_CompactAndExpandMiniBlocksShouldResultTheSameMiniBlocks(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction_test.go]: added some unit tests and moved that test to transactions_test.go (where it belongs)

File: process/block/preprocess/transactions_test.go [hunk 16]
```diff
@@ -4,6 +4,9 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+
+	"github.com/ElrondNetwork/elrond-go/data"
```
[the remark this answers is not in the document]
Comment (bogdan-rosianu) [reply process/block/preprocess/transactions_test.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 6 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 7 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 8 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 9 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 10 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 11 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: done

File: process/block/preprocess/miniBlocksCompaction.go (hunk 13 above)
Comment (bogdan-rosianu) [reply process/block/preprocess/miniBlocksCompaction.go]: renamed to mapMinSenderNonce in order to keep the naming pattern for the other maps in the file

File: process/block/preprocess/miniBlocksCompaction.go [hunk 17]
```diff
@@ -146,6 +132,20 @@ func (mbc *miniBlocksCompaction) calculateUsedGasInMiniblock(miniBlock *block.Mi
...
 }
 
+func (mbc *miniBlocksCompaction) computeGasConsumedByTx(txHash []byte) (uint64, error) {
+	tx, ok := mbc.mapHashToTx[string(txHash)]
+	if !ok {
+		return 0, process.ErrMissingTransaction
+	}
+
+	txGasLimit := mbc.economicsFee.ComputeGasLimit(tx)
+	if isSmartContractAddress(tx.GetRecvAddress()) {
```
Comment (sasurobert) [inline process/block/preprocess/miniBlocksCompaction.go]: this should be done only if destination is not current shard.

Comment (sasurobert) [reply process/block/preprocess/miniBlocksCompaction.go]: you can leave it right now. As this restriction is what we need. Actual gas consumption will be added on a new pull request.

Comment (SebastianMarian) [reply process/block/preprocess/miniBlocksCompaction.go]: ok

**Pull request merged**